pkg/controller/meta_pod_controller.go

```diff
@@ -0,0 +1,231 @@
+func (mpc *MetaPodController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	defaultRequeue2sResult := ctrl.Result{RequeueAfter: time.Second * 2}
+	pendingRequeue10sResult := ctrl.Result{RequeueAfter: time.Second * 10}
+
+	// only reconcile when this is related to a meta pod
+	reqPod := &corev1.Pod{}
+	mpc.Get(ctx, req.NamespacedName, reqPod)
+	rwComponent, ok := reqPod.Labels[consts.LabelRisingWaveComponent]
+	if !ok || rwComponent != consts.ComponentMeta {
+		return defaultRequeue2sResult, nil
+	}
+
+	log := log.FromContext(ctx)
+
+	originalReqPod := reqPod.DeepCopy()
+	oldRole := reqPod.Labels[consts.LabelRisingWaveMetaRole]
+	timeoutCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+
+	podIP, ok := getPodIP(reqPod)
+	if !ok {
+		log.Info("error when reconciling this pod. Assuming that pod is pending")
+		return pendingRequeue10sResult, nil
+	}
+	podPort, ok := getMetaPort(reqPod)
+	if !ok {
+		log.Error(fmt.Errorf("error when reconciling this pod"), "")
+		return defaultRequeue2sResult, nil
+	}
+
+	newRole, ok := mpc.getMetaRole(timeoutCtx, podIP, podPort, reqPod.Name)
+	if !ok {
+		return defaultRequeue2sResult, nil
+	}
+	reqPod.Labels[consts.LabelRisingWaveMetaRole] = newRole
+
+	// only update if something changed
+	if oldRole == newRole {
+		return defaultRequeue2sResult, nil
+	}
+
+	if err := mpc.Patch(ctx, reqPod, client.MergeFrom(originalReqPod)); err != nil {
+		log.Error(err, "unable to update this Pod")
+		return defaultRequeue2sResult, nil
+	}
+
+	// Update other meta components only if we have a change of leadership
+	if oldRole != consts.MetaRoleLeader && newRole != consts.MetaRoleLeader {
+		return defaultRequeue2sResult, nil
+	}
+
+	otherMetaPods, err := mpc.getOtherMetPods(reqPod, ctx)
+	if err != nil {
+		log.Error(err, "unable to retrieve other meta pods")
+		return defaultRequeue2sResult, nil
+	}
+
+	for _, pod := range otherMetaPods {
+		podIP, ok := getPodIP(&pod)
+		if !ok {
+			log.Info("error when reconciling other meta pod %s. Assuming that pod is pending", pod.Name)
+			return pendingRequeue10sResult, nil
+		}
+		podPort, ok := getMetaPort(&pod)
+		if !ok {
+			log.Error(fmt.Errorf("error getting port from other meta pod %s", pod.Name), "")
+			return defaultRequeue2sResult, nil
+		}
+
+		// set meta label
+		originalPod := pod.DeepCopy()
+		oldRole := pod.Labels[consts.LabelRisingWaveMetaRole]
+		timeoutCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
+		defer cancel()
+		newRole, ok := mpc.getMetaRole(timeoutCtx, podIP, podPort, pod.Name)
+		if !ok {
+			return defaultRequeue2sResult, nil
+		}
+		pod.Labels[consts.LabelRisingWaveMetaRole] = newRole
+
+		// only update if something changed
+		if oldRole == newRole {
+			continue
+		}
+
+		if err := mpc.Patch(ctx, &pod, client.MergeFrom(originalPod)); err != nil {
+			log.Error(err, "unable to update Pod")
+			continue
+		}
+		log.Info("updated meta pod", "pod", pod.Name)
```

Comment:
Let's make the label update a function for better readability. Besides, I'm not a big fan of returning the same values everywhere when there are many return points. I prefer to make the inner logic a function and return it after an invocation, .e.g,.

```go
func before() {
	if A {
		return X
	}
	if B {
		return X
	}
	if C {
		return X
	}
	return X
}

func after_inner() {
	if A {
		return
	}
	if B {
		return
	}
	if C {
		return
	}
	return
}

func after() {
	after_inner()
	return X
}
```

Reply:
Hey, I split the function to make it more readable. Let me know if this version is better.